Refactor the POI system and some minor QOL improvements

Molitvan opened the pull request:
The POIGroup class improves the POI system and allows for features like an object tracker and custom way points.
- Refactor the POI system with the POIGroup class

# Changelog
## Feature Updates
- Your pets, other's pets, bosses, players, and vehicles now have their own sounds
- Dropped item detection also now extends to projectiles on the ground which can be picked up like arrows and tridents

## Bug Fixes
- Some neutral mobs are now recognized as hostile when angry like wolves, instead of always being passive

Molitvan committed:
Refactor: Add a new POIGroup class

- Added a new POIGroup class that will be needed for some new features like lockon target switching and custom waypoints
- Made the POI audio cue system use the new POIGroup class
- Locking and marking still depend on the old system, so the code for it is still there for now

Molitvan committed:
Make entity locking and marking use POI groups

- Entity locking and marking now use the new POIGroup class
- Removed some unnecessary comments from the POIGroup class
- Removed legacy code that's not needed anymore

Molitvan committed:
Merge branch 'minecraft-access:1.21' into poi-target-switching

BrailleBennett committed:
Added new POI groups and more sounds

- Your pets group which plays a flute sound
- Other pets group which plays a cow bell sound
- Fish/other under water creatures now count as passive entities
- Bosses group which plays a pling sound for any entity (excluding the iron golem) over 79 health
- Entities which the client can tell are angry now change to a hostile mob once that happens
- Vehicles now plays an iron xylophone sound
- Players now plays a chime sound
- Minor reformatting

BrailleBennett committed:
more minor reformatting in POI entities/blocks

Molitvan committed:
Refactor POIEntities logic to only include the entity in the first group it can belong to

- Rewrote the POIEntities sound playing/group checking logic

Molitvan committed:
Group api improvements

- When using isBlockInGroup and isEntityInGroup, if the subject coresponds to the group, it will be added to it
- Added methods for manually clearing the group and adding new members to it

Molitvan committed:
Make block locking and marking work with the new POIGroup class

Molitvan committed:
Remove legacy POIBlocks code

boholder commented on common/src/main/java/com/github/khanshoaib3/minecraft_access/features/point_of_interest/BlockFilter.java:
```diff
@@ -0,0 +1,10 @@
+import net.minecraft.block.BlockState;
+import net.minecraft.util.math.BlockPos;
+
+// the reason for this existing is Java being weird and not having a built-in functional interface that takes 2 parameters and returns a single value
```
Nope, there is, BiFunction<T,U,R>

boholder commented on common/src/main/java/com/github/khanshoaib3/minecraft_access/features/point_of_interest/LockingHandler.java:
```diff
@@ -25,9 +25,10 @@
 
 /**
- * Locks on to the nearest entity or block.<br><br>
+ * Locks on to the narest entity or block.<br><br>
```
I think "nearest" is the right one?...

boholder commented on common/src/main/java/com/github/khanshoaib3/minecraft_access/features/point_of_interest/POIBlocks.java:
```diff
@@ -128,15 +127,9 @@ public void update(boolean onMarking, Block markedBlock) {
-            otherBlocks = new TreeMap<>();
-            markedBlocks = new TreeMap<>();
+            for (POIGroup group : builtInGroups.values()) {
+                group.clearBlocks();
```
I want to confirm: Have you decided to share detected targets between ticks or cache scanning result for other usage? If not, create and replace map instances will be better for avoiding unwanted timing issues, for example the map isn't cleared timely.

boholder commented on common/src/main/java/com/github/khanshoaib3/minecraft_access/features/point_of_interest/POIBlocks.java:
It's really relaxing seeing POI classes being simpler and clearer... Thanks!

boholder commented on common/src/main/java/com/github/khanshoaib3/minecraft_access/features/point_of_interest/POIGroup.java:
```diff
@@ -0,0 +1,127 @@
+        this.blockFilter = blockFilter;
+    }
+
+    public String getName() {
```
You can get rid of getter methods by announcing class fields as `public final`

boholder commented on common/src/main/java/com/github/khanshoaib3/minecraft_access/features/point_of_interest/POIGroup.java:
```diff
@@ -0,0 +1,127 @@
+        return soundPitch;
+    }
+
+    public void filterEntities(List<Entity> scannedEntities) {
```
Like what I said before, consider assigning new instances to class fields instead of reuse old one, unless you have other purposes that reuse old data.

boholder commented on common/src/main/java/com/github/khanshoaib3/minecraft_access/features/point_of_interest/POIGroup.java:
```diff
@@ -0,0 +1,127 @@
+
+    public boolean isBlockInGroup(BlockState block, BlockPos pos) {
+        boolean result = blockFilter.apply(block, pos);
+        if (result) blocks.put(pos, block);
```
Why a predicate method modifies content of class fields? It's not pure.

Molitvan committed:
Remove BlockFilter and replace it with BiFunction

Molitvan committed:
Fix typo and remove unused import in LockingHandler

Molitvan committed:
Replace getter methods with final variables in POIGroup

Molitvan committed:
Replace .clear with assigning empty Lists/TreeMaps in POIGroup

Molitvan committed:
Renamed isInGroup methods to checkAndAdd

This is in order to follow the rule that predicate methods (those starting with is) shouldn't modify anything and just return true or false

boholder reviewed (dismissed):
Nice.

Molitvan committed:
Add group clearing to POIEntities

Forgot to do this before

BrailleBennett committed:
Enabled block sounds by default

sgmedev committed:
Merge branch '1.21' into poi-target-switching
